Source/WebCore/html/HTMLDetailsElement.cpp

```diff
@@ -160,23 +160,53 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
                 root->appendChild(*m_defaultSlot);
                 queueDetailsToggleEventTask(DetailsState::Closed, DetailsState::Open);
                 if (auto& name = attributeWithoutSynchronization(nameAttr); document().settings().detailsNameAttributeEnabled() && !name.isEmpty()) {
-                    Vector<RefPtr<HTMLDetailsElement>> otherElementsInNameGroup;
-                    for (auto& detailsElement : descendantsOfType<HTMLDetailsElement>(rootNode())) {
-                        if (&detailsElement != this && detailsElement.attributeWithoutSynchronization(nameAttr) == name)
-                            otherElementsInNameGroup.append(&detailsElement);
-                    }
-                    for (RefPtr otherDetailsElement : otherElementsInNameGroup)
+                    bool shouldFireMutationEvents = document().shouldFireMutationEvents();
+                    document().setShouldFireMutationEvents(false);
+                    for (RefPtr otherDetailsElement : otherElementsInNameGroup())
                         otherDetailsElement->removeAttribute(openAttr);
+                    document().setShouldFireMutationEvents(shouldFireMutationEvents);
                 }
             } else {
                 root->removeChild(*m_defaultSlot);
                 queueDetailsToggleEventTask(DetailsState::Open, DetailsState::Closed);
             }
         }
-    } else
+    } else {
+        ensureDetailsExclusivityAfterMutation();
         HTMLElement::attributeChanged(name, oldValue, newValue, attributeModificationReason);
+    }
+}
+
+Node::InsertedIntoAncestorResult HTMLDetailsElement::insertedIntoAncestor(InsertionType insertionType, ContainerNode& parentOfInsertedTree)
+{
+    ensureDetailsExclusivityAfterMutation();
+    return HTMLElement::insertedIntoAncestor(insertionType, parentOfInsertedTree);
+}
+
+Vector<RefPtr<HTMLDetailsElement>> HTMLDetailsElement::otherElementsInNameGroup()
+{
+    Vector<RefPtr<HTMLDetailsElement>> otherElementsInNameGroup;
+    for (auto& detailsElement : descendantsOfType<HTMLDetailsElement>(rootNode())) {
+        if (&detailsElement != this && detailsElement.attributeWithoutSynchronization(nameAttr) == attributeWithoutSynchronization(nameAttr))
+            otherElementsInNameGroup.append(&detailsElement);
+    }
+    return otherElementsInNameGroup;
 }
 
+void HTMLDetailsElement::ensureDetailsExclusivityAfterMutation()
+{
+    if (auto& name = attributeWithoutSynchronization(nameAttr); m_isOpen && !name.isEmpty() && document().settings().detailsNameAttributeEnabled()) {
+        bool shouldFireMutationEvents = document().shouldFireMutationEvents();
+        document().setShouldFireMutationEvents(false);
+        for (RefPtr otherDetailsElement : otherElementsInNameGroup()) {
+            if (otherDetailsElement->m_isOpen) {
+                toggleOpen();
```

Comment:
This seems fragile to me. Are we guaranteed that mutation events are the only way `toggleOpen` can lead to arbitrary execution of JavaScript? Because if anything runs and then in turn does other DOM modification, it will run with mutation events disabled. Calling `setShouldFireMutationEvents(false)` alone seems like a not entirely robust guarantee that no events of any kind will be dispatched. I’m sure that’s true right now, but I can so easily imagine us making a mistake later as we evolve the implementation.